### ISSUE TITLE
recursiveTypeRelatedTo OOM on typescript 3.0.1

### ISSUE THREAD

**Issue by RoboPhred** [Microsoft/TypeScript#26201]
**TypeScript Version:**  3.0.1
NodeJS 10.4.1
Windows 10

**Search Terms:**
recursiveTypeRelatedTo
OOM
Out Of Memory

**Code**

Project at `https://github.com/RoboPhred/oni-duplicity`

**Expected behavior:**
Successful compiles over several changes in watch mode.

**Actual behavior:**

OOM crash after one or two changes.  The location floats around as expected, but quite often tends to fall on `recursiveTypeRelatedTo`:
```

<--- Last few GCs --->

[40596:00000284D9B72D60]   311338 ms: Scavenge 1377.7 (1422.0) -> 1376.8 (1422.5) MB, 3.8 / 0.0 ms  (average mu = 0.176, current mu = 0.120) allocation failure
[40596:00000284D9B72D60]   311348 ms: Scavenge 1377.8 (1422.5) -> 1376.9 (1423.0) MB, 3.1 / 0.0 ms  (average mu = 0.176, current mu = 0.120) allocation failure
[40596:00000284D9B72D60]   311356 ms: Scavenge 1377.8 (1423.0) -> 1376.9 (1423.5) MB, 3.5 / 0.0 ms  (average mu = 0.176, current mu = 0.120) allocation failure


<--- JS stacktrace --->

==== JS stack trace =========================================

    0: ExitFrame [pc: 0000008EA0D841C1]
    1: StubFrame [pc: 0000008EA0D95272]
Security context: 000000FADCE9E589 <JSObject>
    2: recursiveTypeRelatedTo(aka recursiveTypeRelatedTo) [0000016A2698F529] [D:\dev\git\RoboPhred\oni-duplicity\node_modules\typescript\lib\typescript.js:~39655] [pc=0000008EA280772A](this=000003A84DB022E1 <undefined>,source=0000016A2698DCB1 <TypeObject map = 000001905C3C7A09>,target=0000016A269884E9 <TypeObject m...

FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
 1: node::DecodeWrite
 2: node_module_register
 3: v8::internal::FatalProcessOutOfMemory
 4: v8::internal::FatalProcessOutOfMemory
 5: v8::internal::Heap::MaxHeapGrowingFactor
 6: v8::internal::ScavengeJob::operator=
 7: v8::internal::ScavengeJob::operator=
 8: v8::internal::Heap::MaxHeapGrowingFactor
 9: v8::internal::Factory::NewFillerObject
10: v8::internal::ParseInfo::GetFlag
11: 0000008EA0D841C1
```

**Related Issues:** 
- [Compiler OOM in watch mode and with --d ](#21140)
Issue has a similar OOM location, but starts in a typescript version that worked fine for me.

I have a Webpack+React project that has been going strong since typescript 2.8.  However, upon upgrading to typescript 3.0.1, I get regular OOM crashes in the same typescript function while making changes in watch mode.

These crashes have been reproduced on 3 different computers, 

**Reproduction**

Clone the repo at https://github.com/RoboPhred/oni-duplicity.  Set up and run with `npm ci && npm start`.

Add some text to a component.  I chose adding `Test` to the end of line 95 in `src/pages/SaveEditor/components/editors/game-objects/GeyserEditor.tsx`
Save the modified file.  Await (a) successful compile or (b) OOM crash.
If compile succeeded, remove the added content, save, repeat.

I often experience the crash on the very first change/recompile session, and have always received a crash by the second attempt.

**Comment by sheetalkamat** [comment]
@RoboPhred can you please check if this repros with typescript@next. I looked at this (by doing edits multiple times) and everytime this seems to work correctly without OOM exception. Thanks.

**Comment by RoboPhred** [comment]
I just tested it against typescript@next (which pulled `3.1.0-dev.20180830`) and everything is working great!  I cannot reproduce the OOM, and compilation time is significantly faster.

Whatever bug this was, it looks like it got fixed since then.

For reference, I tested it on the v2.1.0 tag of my repo, which is the same commit I originally had the problem on.  The problem still existed on TS 3.0.1 but went away when typescript@next was applied.

Thanks for your attention on this, I appreciate it.
